tests/measurements/test_purity_measurement.py

```diff
@@ -0,0 +1,374 @@
+
+        dev = qml.device(device, wires=2)
+
+        @qml.qnode(dev, interface="tf")
+        def circuit(x):
+            qml.IsingXX(x, wires=[0, 1])
+            return qml.purity(wires=wires)
+
+        purity = circuit(tf.Variable(param))
+        expected_purity = expected_purity_ising_xx(param) if is_partial else 1
+        assert qml.math.allclose(purity, expected_purity)
+
+    @pytest.mark.tf
+    @pytest.mark.parametrize("device", grad_supported_devices)
+    @pytest.mark.parametrize("param", parameters)
+    @pytest.mark.parametrize("wires,is_partial", wires_list)
+    @pytest.mark.parametrize("diff_method", diff_methods)
+    def test_IsingXX_qnode_purity_grad_tf(self, device, param, wires, is_partial, diff_method):
+        """Test purity for a QNode gradient with tf."""
+
+        import tensorflow as tf
+
+        dev = qml.device(device, wires=2)
+
+        @qml.qnode(dev, interface="tf", diff_method=diff_method)
+        def circuit(x):
+            qml.IsingXX(x, wires=[0, 1])
+            return qml.purity(wires=wires)
+
+        grad_expected_purity = expected_purity_grad_ising_xx(param) if is_partial else 0
+
+        param = tf.Variable(param)
+        with tf.GradientTape() as tape:
+            purity = circuit(param)
+
+        grad_purity = tape.gradient(purity, param)
+
+        assert qml.math.allclose(grad_purity, grad_expected_purity, rtol=1e-04, atol=1e-05)
+
+    @pytest.mark.parametrize("device", devices)
+    @pytest.mark.parametrize("return_new", return_new)
+    def test_qnode_entropy_no_custom_wires(self, device, return_new):
+        """Test that purity cannot be returned with custom wires."""
+
+        if return_new:
+            qml.enable_return()
+
+        dev = qml.device(device, wires=["a", 1])
+
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.IsingXX(x, wires=["a", 1])
+            return qml.purity(wires=["a"])
+
+        with pytest.raises(
+            qml.QuantumFunctionError,
+            match="Returning the purity is not supported when using custom wire labels",
+        ):
+            circuit(0.1)
+
+        if return_new:
+            qml.disable_return()
+
+    @pytest.mark.parametrize("return_new", return_new)
+    def test_shot_warnings(self, return_new):
+        """Tests the warning message when using shots"""
+
+        if return_new:
+            qml.enable_return()
+
+        dev = qml.device("default.qubit", wires=2, shots=1000)
+
+        @qml.qnode(device=dev)
+        def circuit(x):
+            qml.IsingXX(x, wires=[0, 1])
+            return qml.purity(wires=[0, 1])
+
+        with pytest.warns(
+            UserWarning,
+            match="Requested measurement PurityMP with finite shots; the returned "
+            "state information is analytic and is unaffected by sampling. To silence "
+            "this warning, set shots=None on the device.",
+        ):
+            circuit(0.5)
+
+        if return_new:
+            qml.disable_return()
```

Comment:
No need to test this. It is already tested for the `StateMeasurement` class.

Reply:
It shouldn't complain given that you didn't add logic for the new return types in the `process` method.

What lines are not covered?

Reply:
Ok I realize that now the warning comes from the general method, so this test can be removed.